### DISCUSSION TITLE
gomplate throw error "map has no entry for key \"spectra\"

### DISCUSSION THREAD

**Discussion by rameshkjha** [hairyhenderson/gomplate#1185] [Q&A] [answered]
I am trying to run gomplate on my overlay kustomization directory and running into an error 
gomplate --input-dir ./spectra/ --output-dir kustomize-config --datasource config=./base/config.yaml 

ERR  error="template: spectra/kustomization.yaml:7:23: executing \"spectra/kustomization.yaml\" at <.spectra.instance.name>: map has no entry for key \"spectra\""

Here is my directory structure and file content. Appreciate any input on what is wrong here ? Thanks in advance !

kustomize
|               | base
|                         config.yaml
|               |spectra
|               |kustomization.yaml

file 1) ./base/config.yaml
spectra:
  instance:
    name: kustomize-dashboardsvc
  stage:
    name: alpha

file 2) ./spectra/kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
  - ../base

namespace: "{{ .spectra.instance.name }}"

commonLabels:
  stage: "{{ .spectra.stage.name }}"

**Comment by hairyhenderson** [comment, marked as answer]
Hi @rameshkjha, thanks for filing this - I've converted it to a discussion since this isn't a bug report or feature request.

The `--datasource` flag defines a datasource that can be accessed with the [`datasource`/`ds`](https://docs.gomplate.ca/functions/data/#datasource) function. You're using it to define a datasource named `config`. So in this case you could access it with something like this:

```
namespace: "{{ (ds "config").spectra.instance.name }}"
```

You can also use [the context](https://docs.gomplate.ca/syntax/#the-context) as an alternative which makes for nicer (IMO) syntax, but there are tradeoffs. For your usage I'd recommend it. To use that you need [the `--context`/`-c` flag](https://docs.gomplate.ca/usage/#context-c):

```console
$ gomplate ... --context config=./base/config.yaml
...
```

Then, you can access the context and the datasources attached to it with `.`:

```
namespace: "{{ .config.spectra.instance.name }}"
```

There's a special-case available as well, most useful if you know you'll only ever have a single datasource on the context, which is to use `.` as the name instead of something like `config`:

```console
$ gomplate ... --context .=./base/config.yaml
...
```

Then you can use syntax like you listed:

```
namespace: "{{ .spectra.instance.name }}"
```

Please read the docs for further information around datasources and the context. Hope this helps!

**Reply by rameshkjha** [reply]
@hairyhenderson 
Thank you very much for explaining it and helping out with various options for my usecase.

Both the combination worked for me.

gomplate --input-dir ./spectra/ --output-dir kustomize-config --context .=./base/config.yaml
namespace: "{{ .spectra.instance.name }}"

gomplate --input-dir ./spectra/ --output-dir kustomize-config --context config=./base/config.yaml 
namespace: "{{ .config.spectra.instance.name }}"

I have just started exploring gomplate and loving it.
Appreciate your help.

**Reply by hairyhenderson** [reply]
@rameshkjha Glad that helped! Can you mark my response as an answer? Thanks 🙂

**Comment by rameshkjha** [comment]
Marking it as answered.
